README.md
# Fluct
Fluct is a framework to build server-less web applications using
[Lambda](https://aws.amazon.com/lambda/) and [API Gateway](https://aws.amazon.com/api-gateway/).
This stack brings about great advantages in the web development,
such as fully isolated components, infinitely scalability, easy and rapid development,
cheap server cost, and more and more.

- [Getting Started](#getting-started)
  - [1. Install](#1-install)
  - [2. Create an application](#2-create-an-application)
  - [3. Create an action](#3-create-an-action)
  - [4. Set up the action](#4-set-up-the-action)
  - [5. Set up the package.json](#5-set-up-the-package-json)
  - [6. Deploy it](#6-deploy-it)
  - [7. Done!](#7-done)
- [CLI](/docs/cli.md)
- [FAQ](/docs/faq.md)
- [Demo](https://github.com/r7kamura/fluct-example)

## Getting Started
### 1. Install
Install `fluct` executable via npm.

```
$ npm install fluct -g
```

### 2. Create an application
Create a new application with an application name.

```
$ fluct new myapp
Created ./myapp
Created ./myapp/.gitignore
Created ./myapp/actions
Created ./myapp/actions/.keep
Created ./myapp/package.json
```

### 3. Create an action
Enter the application folder and generate a new action.

```
$ cd myapp
$ fluct generate list_users
Created ./actions/list_users
Created ./actions/list_users/index.js
Created ./actions/list_users/package.json
```

### 4. Set up the action
Update the action's package.json with proper httpMethod and path.

```
$ vi actions/list_users/package.json
$ cat actions/list_users/package.json
{
  "name": "list_users",
  "private": true,
  "fluct": {
    "contentType": "text/html",
    "httpMethod": "GET",
    "path": "/users",
    "statusCode": 200
  }
}
```

### 5. Set up the package.json
Head over to [AWS Console](https://console.aws.amazon.com) and create a new IAM role
that has `AWSLambdaBasicExecutionRole`, then set its role name and your account ID
to application's package.json (Account ID is written in integers in user's ARN).
This role is used to allow API Gateway to invoke Lambda functions.

```
$ vi package.json
$ cat package.json
{
  "name": "myapp",
  "private": true,
  "fluct": {
    "accountId": "012345678912",
    "restapiId": null,
    "roleName": "fluct-example-role"
  }
}
```

### 6. Deploy it
Deploy your application to Lambda and API Gateway.

```
$ fluct deploy
Created zip file: ./actions/list_users/lambda.zip
Uploaded function: list_users
Updated endpoint: GET /users
Deployed: https://123ge4oabj.execute-api.us-east-1.amazonaws.com/production
```

### 7. Done!
Try to send HTTP requests to your endpoints.

```
$ curl https://123ge4oabj.execute-api.us-east-1.amazonaws.com/production/users -i
HTTP/1.1 200 OK
Content-Type: text/html
Content-Length: 13
Connection: keep-alive
Date: Tue, 11 Aug 2015 19:22:33 GMT
x-amzn-RequestId: 512f8391-405e-11e5-acef-2125b850bbe1
X-Cache: Miss from cloudfront
Via: 1.1 6145a790e7dca1c0c567e1f5decce786.cloudfront.net (CloudFront)
X-Amz-Cf-Id: 5LvHm6SaEQnTj1ubwlCvJhew6G86AU6FFEGB2ic3FI-r7kwNfwDCXg==

Hello, world!
```

Now that you’re up and running, here are a few things you should know.
See [CLI](/docs/cli.md) to use utility commands
to develop your application, and see [FAQ](/docs/faq.md) to know how to customize your application.


CHANGELOG.md
## v0.4.2
- Support ./bin/* files (#48)

## v0.4.1
- Add some dependencies

## v0.4.0
- Use aws-sdk-promise

## v0.3.6
- Escape input in request template (#33)
- Support memorySize option (#34)

## v0.3.4
- Support configuration of AWS region (#31)

## v0.3.3
- Fix missing lib dir in npm package

## v0.3.2
- Support node-config

## v0.3.1
- Fix fluct-deploy bug

## v0.3.0
- Add new command: `fluct open`
- Require accountID and role name
- Only include production dependencies into lambda.zip
- Support other response status code

## v0.2.7
- Include HTTP request data into event object

## v0.2.6
- Support any Content-Type

## v0.2.5
- Fix `fluct deploy` error on uploading new function

## v0.2.4
- Fix function upload bug

## v0.2.3
- Change `fluct generate` command interface

## v0.2.2
- Use common package.json in all actions
- Add `fluct deployments` command

## v0.2.1
- Add missing dependency (#14)
- Fix `fluct deploy` bug

## v0.2.0
- Change action directory structure
- Include node_modules into zip file for Lambda function
- Improve `fluct deploy` logs
- Improve `fluct routes` format

## v0.1.2
- Fix broken `fluct server` command

## v0.1.1
- Add new command: `fluct install`
- Provide command aliases

## v0.1.0
- Use text/html response by default

## v0.0.8
- Add new command: `fluct routes`
- Create deployment at the end of `fluct deploy`

## v0.0.7
- Create restapi only at 1st time
- Improve generated package.json content

## v0.0.6
- Add new command: `fluct deploy`

## v0.0.5
- Make handler reloadable
- Fix missing templates dir

## v0.0.4
- Add new command: `fluct server`
- Fix package.json bin property

## v0.0.3
- Add missing bin dir into fluct npm package

## v0.0.2
- Add new command: `fluct new`
- Add new command: `fluct generate action`

## v0.0.1
- 1st Release


src/action.js
import Application from './application'
import fs from 'fs'

/**
 * @class
 */
export default class Action {
  /**
   * @param {String} name
   */
  constructor({ application, name }) {
    this.application = application;
    this.name = name;
  }

  /**
   * @return {String}
   * @example 'text/html'
   */
  getContentType() {
    return this.getMetadata().contentType;
  }

  /**
   * @return {String}
   */
  getDirectoryPath() {
    return `./actions/${this.name}`;
  }

  /**
   * @return {String}
   */
  getFunctionArn() {
    return `arn:aws:lambda:${this.application.getRegion()}:${this.application.getAccountId()}:function:${this.getName()}`;
  }

  /**
   * @return {Function}
   */
  getHandler() {
    const handlerScriptPath = `${process.cwd()}/${this.getDirectoryPath()}`;
    delete(require.cache[handlerScriptPath]);
    return require(handlerScriptPath).handler;
  }

  /**
   * @return {String}
   */
  getHandlerId() {
    return 'index.handler';
  }

  /**
   * @return {String}
   */
  getHttpMethod() {
    return this.getMetadata().httpMethod;
  }

  /**
   * @return {String}
   */
  getName() {
    return this.getMetadata().name;
  }

  /**
   * @return {Object}
   */
  getMetadata() {
    return JSON.parse(
      fs.readFileSync(`${this.getDirectoryPath()}/fluct.json`)
    );
  }

  /**
   * @return {String}
   */
  getPath() {
    return this.getMetadata().path;
  }

  /**
   * @return {Object}
   */
  getRequestTemplates() {
    return {
      'application/json': `
      {
        "accountId": "$context.identity.accountId",
        "apiId": "$context.apiId",
        "apiKey": "$context.identity.apiKey",
        "caller": "$context.identity.caller",
        "headers": {
      #foreach( $key in $input.params().header.keySet() )
          "$key": "$util.escapeJavaScript($input.params().header.get($key))"#if( $foreach.hasNext ),#end
      #end
        },
        "httpMethod": "$context.httpMethod",
        "path": "$context.resourcePath",
        "pathParameters": {
      #foreach( $key in $input.params().path.keySet() )
          "$key": "$util.escapeJavaScript($input.params().path.get($key))"#if( $foreach.hasNext ),#end
      #end
        },
        "queryParameters": {
      #foreach( $key in $input.params().querystring.keySet() )
          "$key": "$util.escapeJavaScript($input.params().querystring.get($key))"#if( $foreach.hasNext ),#end
      #end
        },
        "requestId": "$context.requestId",
        "requestParameters": $input.json('$'),
        "resourceId": "$context.resourceId",
        "sourceIp": "$context.identity.sourceIp",
        "stage": "$context.stage",
        "user": "$context.identity.user",
        "userAgent": "$context.identity.userAgent",
        "userArn": "$context.identity.userArn"
      }`.replace(/^      /gm, '')
    };
  }

  /**
   * @return {Object}
   * @example { 'text/html': 'Empty' }
   */
  getResponseModels() {
    const contentType = this.getContentType();
    const returnedValue = {};
    if (contentType !== 'application/json') {
      returnedValue[contentType] = 'Empty';
    }
    return returnedValue;
  }

  /**
   * @return {Object}
   */
  getResponseTemplates() {
    if (this.getMetadata().contentType == 'application/json') {
      return {}
    } else {
      return { 'text/html': "$input.path('$')" }
    }
  }

  /**
   * @return {Integer}
   */
  getStatusCode() {
    return this.getMetadata().statusCode || 200;
  }

  /**
   * @return {Integer}
   */
  getTimeout() {
    return 60;
  }

  /**
   * @return {Integer}
   */
  getMemorySize() {
    return this.getMetadata().memorySize || 128;
  }

  /**
   * @return {String}
   */
  getUri() {
    return `arn:aws:apigateway:${this.application.getRegion()}:lambda:path/2015-03-31/functions/${this.getFunctionArn()}/invocations`;
  }

  /**
   * @return {String}
   */
  getZipPath() {
    return `${this.getDirectoryPath()}/lambda.zip`;
  }

  /**
   * @param {http.IncomingMessage} request
   * @param {http.ServerResponse} response
   */
  run(request, response) {
    this.getHandler()(
      {},
      {
        succeed: (value) => {
          response.header('Content-Type', this.getContentType());
          response.send(value);
        }
      }
    );
  }
}
